src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs

```diff
@@ -0,0 +1,293 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    /// <summary>
+    /// A LIFO semaphore.
+    /// Waits on this semaphore are uninterruptible.
+    /// </summary>
+    internal partial class LowLevelLifoSemaphore : IDisposable
+    {
+        private CacheLineSeparatedCounts _separated;
+
+        private int _maximumSignalCount;
+
+        private static int s_processorCount = Environment.ProcessorCount;
+
+        public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount)
+        {
+            Debug.Assert(initialSignalCount >= 0);
+            Debug.Assert(initialSignalCount <= maximumSignalCount);
+
+            _separated = new CacheLineSeparatedCounts();
+            _separated._counts._signalCount = (uint)initialSignalCount;
+            _maximumSignalCount = maximumSignalCount;
+
+            Create(maximumSignalCount);
+        }
+
+        public bool Wait(int timeoutMs)
+        {
+            // Try to acquire the semaphore or
+            // a) register as a spinner if spinCount > 0 and timeoutMs > 0
+            // b) register as a waiter if there's already too many spinners or spinCount == 0 and timeoutMs > 0
+            // c) bail out if timeoutMs == 0 and return false
+            Counts counts = Counts.VolatileRead(ref _separated._counts);
+            while (true)
+            {
+                Debug.Assert(counts._signalCount <= _maximumSignalCount);
+                Counts newCounts = counts;
+
+                if (counts._signalCount != 0)
+                {
+                    newCounts._signalCount--;
+                }
+                else if (timeoutMs != 0)
+                {
+                    if (LowLevelSpinWaiter.SpinCount > 0 && newCounts._spinnerCount < byte.MaxValue)
```

Comment:
The equivalent of `YieldProcessorWithBackOffNormalized()`, which CoreCLR's `ClrLifoSemaphore` uses, is effectively what `SpinWait.SpinOnce` does. The normalization stuff has not been ported to CoreRT yet, but once it's ported the spin logic in `SpinWait.SpinOnce` would automatically be normalized. It would be great if you could also copy the logic involving `OptimalMaxSpinWaitsPerSpinIteration` from `SpinWait` into `LowLevelSpinWaiter` in place of `RuntimeThread.SpinWait(1 << spinIndex)`.

Reply:
Would you be fine with doing that as part of separate PR? I agree it should be done and it's one of the reasons why I tried to reuse `LowLevelSpinWaiter.Wait` from the new code. I just don't feel good about making too many changes at once and bloating the PR into unreviewable proportions.

Reply:
I tried to take a spin on it, but the algorithm used by `LowLevelSpinWaiter` is different from the spinning logic in `CLRLifoSemaphore`.

`LowLevelSpinWaiter` does couple of spins and then only Yields/Sleep(0)s and never spins again. It will never hit the threshold for `OptimalMaxSpinWaitsPerSpinIteration`. 

`CLRLifoSemaphore` does a few iterations of spinning and then interleaves Sleep(0) with more spinning. `CLRLifoSemaphore` is also very unoptimal on uniprocessor systems since it does meaningless spinning there while only yielding the time slice to other threads makes sense.

Using a straightforward port of the `CLRLifoSemaphore` logic directly in `LowLevelLifoSemaphore` actually hurt the performance significantly for the benchmark linked above:

|                            Method | Toolchain | maxWIC |        Mean |      Error |     StdDev |      Median |
|---------------------------------- |---------- |----------------- |------------:|-----------:|-----------:|------------:|
|     **Burst** |      **Base** |                **1** | **1,987.72 ms** | **39.1397 ms** |  **70.577 ms** | **1,982.95 ms** |
|     Burst |        PR |                1 | 1,134.64 ms | 50.6775 ms | 147.829 ms | 1,120.62 ms |
|     **Burst** |      **Base** |               **10** |   **261.78 ms** |  **7.9853 ms** |  **23.420 ms** |   **257.99 ms** |
|     Burst |        PR |               10 |   182.66 ms |  7.5282 ms |  21.600 ms |   176.22 ms |
|     **Burst** |      **Base** |              **100** |    **55.21 ms** |  **3.4796 ms** |  **10.260 ms** |    **52.13 ms** |
|     Burst |        PR |              100 |    74.95 ms | 10.9991 ms |  32.431 ms |    71.72 ms |
| **Sustained** |      **Base** |                **?** |    **23.52 ms** |  **0.4692 ms** |   **1.361 ms** |    **23.59 ms** |
| Sustained |        PR |                ? |    22.51 ms |  0.5957 ms |   1.738 ms |    21.71 ms |

Notably the StdDev spiked out of charts...

The code I used:
```c#
// TODO: Unify this with LowLevelSpinWaiter
if (s_processorCount > 1)
{
    if (spinIndex < SpinSleep0Threshold || (spinIndex - SpinSleep0Threshold) % 2 != 0)
    {
        // Matches algorithm in SpinWait.SpinOnce
        int n = RuntimeThread.OptimalMaxSpinWaitsPerSpinIteration;
        if (spinIndex <= 30 && (1 << spinIndex) < n)
        {
            n = 1 << spinIndex;
        }
        RuntimeThread.SpinWait(n);
    }
    else
    {
        RuntimeThread.UninterruptibleSleep0();
    }
    
    spinIndex++;
}
else
{
    // On uniprocessor system we only Yield because there cannot be other
    // threads executing in parallel and satisfying the condition. We also
    // compensate for the unnecessary spins by incrementing spinIndex by 2
    // and doing the same number of Yields as in the multi-processor code
    // above.
    RuntimeThread.Yield();
    spinIndex += 2;
}
```

Reply:
Looks like it could have been a flaw in my testing, will re-run the benchmark after verifying that the correct builds are used and that no other process is hogging the machine.

Reply:
Turns I out that the benchmark was really running outdated code and the fluctuation were due to some other process taking the CPU time.

Unfortunately I cannot re-run the benchmarks now. I updated the baseline to be GIT master today and the benchmarking infrastructure got broken:

```
EETypeRva:0x03666B38(System.Reflection.MissingRuntimeArtifactException): This object cannot be invoked because it was metadata-enabled for browsing only: 'System.Diagnostics.DebuggableAttribute.IsJITTrackingEnabled' For more information, please visit  http://go.microsoft.com/fwlink/?LinkID=616867
   at Internal.Reflection.Core.Execution.ExecutionEnvironment.GetMethodInvoker(RuntimeTypeInfo, QMethodDefinition, RuntimeTypeInfo[], MemberInfo) + 0x15d
   at System.Reflection.Runtime.MethodInfos.NativeFormat.NativeFormatMethodCommon.GetUncachedMethodInvoker(RuntimeTypeInfo[], MemberInfo) + 0x3f
   at System.Reflection.Runtime.PropertyInfos.RuntimePropertyInfo.GetValue(Object, BindingFlags, Binder, Object[], CultureInfo) + 0x6c
   at BenchmarkDotNet.Extensions.AssemblyExtensions.Read(DebuggableAttribute, String) + 0x66
   at BenchmarkDotNet.Portability.RuntimeInformation.GetConfiguration() + 0x2c
   at BenchmarkDotNet.Environments.BenchmarkEnvironmentInfo..ctor() + 0x32
   at BenchmarkDotNet.Autogenerated.Runnable_3.Run(IHost, String) + 0x7a
   at BenchmarkDotNet.Autogenerated.UniqueProgramName.AfterAssemblyLoadingAttached(String[]) + 0x194
```

I suspect that this is result of https://github.com/dotnet/corert/pull/6820. @MichalStrehovsky can you help on this one please?

Reply:
I reverted the commit from #6820 and it works again. No amount of rd.xml trickery worked for me (so I probably did it wrong).

Here's the corrected benchmark for the above code:

|                            Method | Toolchain | WIC |        Mean |      Error |     StdDev |      Median |
|---------------------------------- |---------- |----------------- |------------:|-----------:|-----------:|------------:|
|     **Burst** |      **Base** |                **1** | **1,428.26 ms** | **31.2428 ms** | **66.5810 ms** | **1,408.86 ms** |
|     Burst |        PR |                1 |   861.98 ms | 17.1336 ms | 25.6448 ms |   854.40 ms |
|     **Burst** |      **Base** |               **10** |   **240.32 ms** |  **4.8021 ms** |  **8.6593 ms** |   **239.42 ms** |
|     Burst |        PR |               10 |   125.65 ms |  2.4182 ms |  2.4833 ms |   126.34 ms |
|     **Burst** |      **Base** |              **100** |    **44.68 ms** |  **0.8792 ms** |  **1.1432 ms** |    **44.86 ms** |
|     Burst |        PR |              100 |    35.57 ms |  0.7081 ms |  1.4305 ms |    36.16 ms |
| **Sustained** |      **Base** |                **?** |    **20.33 ms** |  **0.2658 ms** |  **0.2356 ms** |    **20.43 ms** |
| Sustained |        PR |                ? |    21.69 ms |  0.3061 ms |  0.2863 ms |    21.61 ms |

Reply:
> I suspect that this is result of https://github.com/dotnet/corert/pull/6820. @MichalStrehovsky can you help on this one please?

Yeah, it is. Sorry about that. The code is actually doing what it's supposed to. The problem is with [this code](https://github.com/dotnet/BenchmarkDotNet/blob/a76f438a81e3e0235a2e9b1249a5280038189689/src/BenchmarkDotNet/Extensions/AssemblyExtensions.cs#L28-L38) in Benchmark.Net.

The newly added default CoreRT compiler switch [IlcGenerateCompleteTypeMetadata](https://github.com/dotnet/corert/blob/master/Documentation/using-corert/optimizing-corert.md) is making sure that metadata is generated even for stuff we didn't compile. In this case we didn't compile the `IsJITTrackingEnabled` property. Previously, this would return `null` from the `GetProperty` call, which is handled, but very much not correct. We now return a `PropertyInfo` that is unusable to help pinpoint such problematic cases (most code out there doesn't actually handle the `null` cases and will just AV in a random location).

There are two possible fixes:
* add `<IlcGenerateCompleteTypeMetadata>false</IlcGenerateCompleteTypeMetadata>` to your project to opt out of the helper, or
* add an RD.XML with this in it:
```xml
<Directives>
    <Application>
        <Assembly Name="System.Runtime">
            <Type Name="System.Diagnostics.DebuggableAttribute" Dynamic="Required All" />
        </Assembly>
    </Application>
</Directives>
```

I'll submit a pull request to Benchmark.net to fix this.

Cc @adamsitnik

Reply:
@MichalStrehovsky Thanks a lot for analyzing it!

Reply:
@filipnavara, your code looks ok to me, this is a bit closer to what I was thinking:

`ClrLifoSemaphore::Wait`:

```c#
    const int Sleep0Threshold = 10;
    for (int spinIndex = 0; spinIndex < spinCount; ++spinIndex)
    {
        // Wait
        //
        // (spinIndex - Sleep0Threshold) % 2 != 0: The purpose of this check is to interleave Thread.Yield/Sleep(0) with
        // Thread.SpinWait. Otherwise, the following issues occur:
        //   - When there are no threads to switch to, Yield and Sleep(0) become no-op and it turns the spin loop into a
        //     busy-spin that may quickly reach the max spin count and cause the thread to enter a wait state. Completing the
        //     spin loop too early can cause excessive context switcing from the wait.
        //   - If there are multiple threads doing Yield and Sleep(0) (typically from the same spin loop due to contention),
        //     they may switch between one another, delaying work that can make progress.
        if (spinIndex < Sleep0Threshold || (spinIndex - Sleep0Threshold) % 2 != 0)
        {
            LowLevelSpinWaiter.Wait(spinIndex, Sleep0Threshold);
        }
        else
        {
            // Not doing RuntimeThread.Yield(), it does not seem to have any benefit over Sleep(0)
            RuntimeThread.UninterruptibleSleep0();
        }

        // Try to acquire the semaphore and unregister as a spinner
        ...
    }
```

`LowLevelSpinWaiter.Wait`:

```c#
        public static void Wait(int spinIndex, int sleep0Threshold)
        {
            Debug.Assert(spinIndex >= 0);
            Debug.Assert(sleep0Threshold >= 0);

            if (spinIndex < sleep0Threshold)
            {
                // Cap the maximum spin count to a value such that many thousands of CPU cycles would not be wasted doing
                // the equivalent of YieldProcessor(), as that that point SwitchToThread/Sleep(0) are more likely to be able to
                // allow other useful work to run. Long YieldProcessor() loops can help to reduce contention, but Sleep(1) is
                // usually better for that.
                //
                // RuntimeThread.OptimalMaxSpinWaitsPerSpinIteration:
                //   - See Thread::InitializeYieldProcessorNormalized(), which describes and calculates this value.
                //
                int n = RuntimeThread.OptimalMaxSpinWaitsPerSpinIteration;
                if (spinIndex <= 30 && (1 << spinIndex) < n)
                {
                    n = 1 << spinIndex;
                }
                RuntimeThread.SpinWait(n);
                return;
            }

            /// <see cref="RuntimeThread.Sleep(int)"/> is interruptible. The current operation may not allow thread interrupt
            /// (for instance, <see cref="LowLevelLock.Acquire"/> as part of <see cref="EventWaitHandle.Set"/>). Use the
            /// uninterruptible version of Sleep(0). Not doing <see cref="RuntimeThread.Yield"/>, it does not seem to have any
            /// benefit over Sleep(0).
            RuntimeThread.UninterruptibleSleep0();

            // Don't want to Sleep(1) in this spin wait:
            //   - Don't want to spin for that long, since a proper wait will follow when the spin wait fails
            //   - Sleep(1) would put the thread into a wait state, and a proper wait will follow when the spin wait fails
            //     anyway (the intended use for this class), so it's preferable to put the thread into the proper wait state
        }
```

> `LowLevelSpinWaiter` does couple of spins and then only Yields/Sleep(0)s and never spins again. It will never hit the threshold for `OptimalMaxSpinWaitsPerSpinIteration`.

Typically, `OptimalMaxSpinWaitsPerSpinIteration` is around 7, so it would reach the threshold at `spinIndex >= 3`. `LowLevelSpinWaiter` does a very tiny bit of spinning so it doesn't really make much difference there.

Reply:
Actually updating my suggestion, sec

Reply:
@filipnavara, your code looks ok to me, this is a bit closer to what I was thinking:

`ClrLifoSemaphore::Wait`:

```c#
    const int Sleep0Threshold = 10;
    for (int spinIndex = 0; spinIndex < spinCount; ++spinIndex)
    {
        LowLevelSpinWaiter.Wait(spinIndex, Sleep0Threshold);

        // Try to acquire the semaphore and unregister as a spinner
        ...
    }
```

`LowLevelSpinWaiter.Wait`:

```c#
        public static void Wait(int spinIndex, int sleep0Threshold)
        {
            Debug.Assert(spinIndex >= 0);
            Debug.Assert(sleep0Threshold >= 0);

            // Wait
            //
            // (spinIndex - Sleep0Threshold) % 2 != 0: The purpose of this check is to interleave Thread.Yield/Sleep(0) with
            // Thread.SpinWait. Otherwise, the following issues occur:
            //   - When there are no threads to switch to, Yield and Sleep(0) become no-op and it turns the spin loop into a
            //     busy-spin that may quickly reach the max spin count and cause the thread to enter a wait state. Completing the
            //     spin loop too early can cause excessive context switcing from the wait.
            //   - If there are multiple threads doing Yield and Sleep(0) (typically from the same spin loop due to contention),
            //     they may switch between one another, delaying work that can make progress.
            if (spinIndex < sleep0Threshold || (spinIndex - sleep0Threshold) % 2 != 0)
            {
                // Cap the maximum spin count to a value such that many thousands of CPU cycles would not be wasted doing
                // the equivalent of YieldProcessor(), as that that point SwitchToThread/Sleep(0) are more likely to be able to
                // allow other useful work to run. Long YieldProcessor() loops can help to reduce contention, but Sleep(1) is
                // usually better for that.
                //
                // RuntimeThread.OptimalMaxSpinWaitsPerSpinIteration:
                //   - See Thread::InitializeYieldProcessorNormalized(), which describes and calculates this value.
                //
                int n = RuntimeThread.OptimalMaxSpinWaitsPerSpinIteration;
                if (spinIndex <= 30 && (1 << spinIndex) < n)
                {
                    n = 1 << spinIndex;
                }
                RuntimeThread.SpinWait(n);
                return;
            }

            /// <see cref="RuntimeThread.Sleep(int)"/> is interruptible. The current operation may not allow thread interrupt
            /// (for instance, <see cref="LowLevelLock.Acquire"/> as part of <see cref="EventWaitHandle.Set"/>). Use the
            /// uninterruptible version of Sleep(0). Not doing <see cref="RuntimeThread.Yield"/>, it does not seem to have any
            /// benefit over Sleep(0).
            RuntimeThread.UninterruptibleSleep0();

            // Don't want to Sleep(1) in this spin wait:
            //   - Don't want to spin for that long, since a proper wait will follow when the spin wait fails
            //   - Sleep(1) would put the thread into a wait state, and a proper wait will follow when the spin wait fails
            //     anyway (the intended use for this class), so it's preferable to put the thread into the proper wait state
        }
```

> `LowLevelSpinWaiter` does couple of spins and then only Yields/Sleep(0)s and never spins again. It will never hit the threshold for `OptimalMaxSpinWaitsPerSpinIteration`.

Typically, `OptimalMaxSpinWaitsPerSpinIteration` is around 7, so it would reach the threshold at `spinIndex >= 3`. `LowLevelSpinWaiter` does a very tiny bit of spinning so it doesn't really make much difference there.

Reply:
Yout suggestion looks good to me, will just need to fix the uni-processor code paths not to spin and yield only. I'll give it a try and if it doesn't break my benchmarks I will use it. Thanks!

Reply:
@MichalStrehovsky big thanks for all the help with the fix!

@filipnavara we have a new nuget package of BenchmarkDotNet with the fix, the version is `0.11.3.995`

You can download it from our CI feed:

```
<packageSources>
  <add key="bdn-CI" value="https://ci.appveyor.com/nuget/benchmarkdotnet" />
</packageSources>
```